core/vm/instructions.go

```diff
@@ -440,13 +442,19 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 	}
 	var upper, lower uint64
 	upper = interpreter.evm.Context.BlockNumber.Uint64()
-	if upper < 2 {
+	if upper < 257 {
 		lower = 0
 	} else {
-		lower = upper - 1
+		lower = upper - 256
 	}
 	if num64 >= lower && num64 < upper {
-		num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+		chainId := interpreter.evm.ChainConfig().ChainID
+		hasher := sha3.NewLegacyKeccak256()
+		hasher.Write(chainId.Bytes())
+		num64Buf := make([]byte, 8)
+		binary.BigEndian.PutUint64(num64Buf, num64)
```

Comment:
We could also encode `num64` using the minimum number of bytes necessary. Currently it's going to be padded to 8 bytes.

I think the current approach is reasonable, just wanted to highlight this. cc @icemelon

Reply:
When do we need to encode? On Zk-EVM side?

Reply:
By encoding I mean this line

```go
binary.BigEndian.PutUint64(num64Buf, num64)
```